### PR TITLE
Use Turborepo for tests and add turbo cache for CI eslint and unit-tests

### LINKED ISSUE #3351: Use Turborepo for tests
**What would you like to be added**:
The same way we use Turborepo to lint our project, we should do the same with tests.

**Why would you like this to be added**:
Speed up tests by using turbo's cache feature.

### PR DESCRIPTION
**Pull request by menghif**
## Issue This PR Addresses
Fixes #3351 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

This PR changes `pnpm test` to use Turborepo and adds support for Turborepo's cache in CI eslint and unit-tests.

The downside of this is not seeing all the green check marks when running tests locally, so I have left the option to run tests without Turborepo by running `pnpm jest`.

For CI I am using a GitHub Action that builds a local server (on port 9080) where Turborepo goes to look for the cache. To make it work, a `TURBO_SERVER_TOKEN` needs to be added to the repo.

If this is successful, we will be able to do the same for our e2e tests.

Resources:
[Custom Remote Cache](https://turborepo.org/docs/features/remote-caching#custom-remote-caches)
[TurboRepo with github artifacts ](https://github.com/marketplace/actions/turborepo-with-github-artifacts)
## Steps to test the PR

`pnpm test`

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: .github/workflows/eslint-ci.yml [hunk 1]
```diff
@@ -14,7 +14,14 @@ jobs:
       - uses: actions/setup-node@v2.5.1
         with:
           cache: 'pnpm'
+
+      - name: Turborepo local server
+        uses: felixmosh/turborepo-gh-artifacts@v1
+        with:
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
+          server-token: ${{ secrets.TURBO_SERVER_TOKEN }}
```
Comment (humphd) [inline .github/workflows/eslint-ci.yml:22]: What is this token supposed to look like?  I'm reading the docs and I don't totally understand the point of it.

**Review by humphd — COMMENTED** [review]
This is awesome.

Should we do a follow-up to run a turbo-repo cache server on staging for local devs to leverage too?  Or is this not needed locally?

File: .github/workflows/eslint-ci.yml (hunk 1 above)
Comment (humphd) [reply .github/workflows/eslint-ci.yml:22]: I've added `TURBO_SERVER_TOKEN` to the repo.

**Comment by menghif** [comment]
> This is awesome.
> 
> Should we do a follow-up to run a turbo-repo cache server on staging for local devs to leverage too? Or is this not needed locally?

I don't think it's needed for local dev since Turborepo keeps a local cache in a `.turbo` folder.

Also there is a new feature in version 1.2 to add signature verification:
https://turborepo.org/blog/turbo-1-2-0#cache-outputs-integrity-and-signature-verification
Is this something we should add?

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by humphd** [comment]
> > This is awesome.
> > Should we do a follow-up to run a turbo-repo cache server on staging for local devs to leverage too? Or is this not needed locally?
> 
> I don't think it's needed for local dev since Turborepo keeps a local cache in a `.turbo` folder.
> 
> Also there is a new feature in version 1.2 to add signature verification: https://turborepo.org/blog/turbo-1-2-0#cache-outputs-integrity-and-signature-verification Is this something we should add?

We're not uploading them to a cache, right?  I'm not sure we need this.

**Comment by menghif** [comment]
> > > This is awesome.
> > > Should we do a follow-up to run a turbo-repo cache server on staging for local devs to leverage too? Or is this not needed locally?
> > 
> > 
> > I don't think it's needed for local dev since Turborepo keeps a local cache in a `.turbo` folder.
> > Also there is a new feature in version 1.2 to add signature verification: https://turborepo.org/blog/turbo-1-2-0#cache-outputs-integrity-and-signature-verification Is this something we should add?
> 
> We're not uploading them to a cache, right? I'm not sure we need this.

Right, it would only be useful if we used an actual server.

File: .github/workflows/eslint-ci.yml [hunk 2]
```diff
@@ -2,6 +2,13 @@ name: ESLint Workflow
 
 on:
   workflow_call:
+    secrets:
+    github-token:
```
Comment (humphd) [inline .github/workflows/eslint-ci.yml:6]: Can you bump these in one-level?

```yml
workflow_call:
  secrets:
    github-token:
      ...
    turborepo-token:
      ...
```

File: .github/workflows/unit-tests-ci.yml [hunk 3]
```diff
@@ -2,6 +2,13 @@ name: Unit Tests Workflow
 
 on:
   workflow_call:
+    secrets:
+    github-token:
```
Comment (humphd) [inline .github/workflows/unit-tests-ci.yml:6]: Same here, bump these in.

**Review by humphd — CHANGES_REQUESTED** [review]

File: .github/workflows/unit-tests-ci.yml (hunk 3 above)
Comment (menghif) [reply .github/workflows/unit-tests-ci.yml]: Just did. I don't understand why CI fails to start, I followed your docker example.

Comment (humphd) [reply .github/workflows/unit-tests-ci.yml]: > [Invalid workflow file: .github/workflows/continuous-integration.yml#L22](https://github.com/Seneca-CDOT/telescope/actions/runs/2150363645/workflow)
The workflow is not valid. .github/workflows/continuous-integration.yml (Line: 22, Col: 21): Invalid secret, github-token is not defined in the referenced workflow. .github/workflows/continuous-integration.yml (Line: 23, Col: 24): Invalid secret, turborepo-token is not defined in the referenced workflow.

From https://github.com/Seneca-CDOT/telescope/actions/runs/2150363645

**Review by humphd — APPROVED** [review]
This looks right to me.  NOTE: we don't run linked workflows from PRs, only what's on `master`, so we have to land this to see it happen.

File: .github/workflows/continuous-integration.yml [hunk 4]
```diff
@@ -18,10 +18,16 @@ jobs:
   # Make sure eslint passes
   eslint:
     uses: Seneca-CDOT/telescope/.github/workflows/eslint-ci.yml@master
+    secrets:
+      github-token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (humphd) [inline .github/workflows/continuous-integration.yml:22]: I think we can't pass this in (it's a synthesized token, not a real one), see https://github.community/t/how-to-pass-github-token-to-reusable-workflows/204905/6.  What if you use `{{ github.token }}` below instead and get rid of this.

**Review by humphd — CHANGES_REQUESTED** [review]
I think it's uphappy with the `GITHUB_TOKEN` stuff

**Review by DukeManh — DISMISSED** [review]

**Review by humphd — APPROVED** [review]

**Pull request merged**